File: apps/web/src/lib/api/system.ts
import { apiFetch } from "./core"

export interface VersionInfo {
  current: string
  /** "stable" (cut from a release tag), "edge" (built from main), or "dev". */
  channel: string
  latest: string
  update_available: boolean
  release_url: string
  /** When GitHub was last asked. */
  checked_at?: string
}

export interface ExposedPort {
  port: number
  service: string
  /** Why this port matters, where the service name alone does not say. */
  note?: string
}

export interface Exposure {
  /** What install.sh saw on the host: none, ufw, firewalld, or unknown. */
  firewall_state: string
  /** When the installer looked (RFC3339). Absent on older installs. */
  checked_at?: string
  /** Populated only when firewall_state is "none". */
  ports: ExposedPort[]
  dismissed: boolean
  /** How this gateway gets certificates: "delegation" when it runs its own
   *  authoritative DNS and holds a wildcard, "ondemand" when DNS stays with
   *  the operator's provider. An internal route differs between the two. */
  dns_mode?: "delegation" | "ondemand"
}

/** A stage of the migration, in the order they run. */
export type MigrationStage =
  | "detect" | "plan" | "credential" | "prepare" | "move" | "cutover" | "rollback" | "finish"

/** One group of the plan and where it has got to. */
export interface GroupProgress {
  id: string
  name: string
  members?: string[]
  moved: boolean
  moved_at?: string
  /** Why the last attempt stopped. The group put itself back, so this is a
   *  reason to read rather than damage to repair. */
  error?: string
  can_move: boolean
  blockers?: string[]
  /** The hostnames this group serves. A group that cannot move loses them when
   *  the ports change hands. */
  domains?: string[]
  data?: string[]
  downtime?: string
}

/** Where the migration has got to, summarised on the host from its journal. */
export interface MigrationStatus {
  updated_at: string
  prepared: boolean
  cut_over: boolean
  finished: boolean
  groups: GroupProgress[]
}

/** A request the console queued for the host agent. */
export interface HostRequestState {
  id: string
  state: "queued" | "running" | "succeeded" | "failed"
  requested_at: string
  finished_at?: string
  error?: string
}

export interface MigrationState {
  agent_reporting: boolean
  detect: unknown | null
  detect_at?: string
  plan: unknown | null
  plan_at?: string
  prepare: unknown | null
  prepare_at?: string
  move: unknown | null
  move_at?: string
  cutover: unknown | null
  cutover_at?: string
  rollback: unknown | null
  rollback_at?: string
  finish: unknown | null
  finish_at?: string
  status: MigrationStatus | null
  status_at?: string
  requests: Record<string, HostRequestState>
}

/** Whether the gateway's host agent is reporting. */
export interface HostAgentStatus {
  reporting: boolean
  version?: string
  started_at?: string
  heartbeat_at?: string
  tasks?: Record<string, { ok: boolean; error?: string; at: string }>
  /** ufw, firewalld, iptables or none. */
  firewall?: string
}

/** Stable slug for the host-exposure advisory. */
export const NOTICE_HOST_EXPOSURE = "host-exposure"

/** The overview's "start here" panel. It hides itself once a service has a
 *  route, so dismissing it is only for someone who would rather not read it. */
export const NOTICE_GETTING_STARTED = "getting-started"

/**
 * Upgrading the server from the console. The API only queues a request; a
 * systemd unit on the gateway runs the upgrade and reports progress back.
 */
export interface UpgradeStatus {
  /** The updater is on (`sudo meshploy updater start`) and the host agent is reporting. */
  enabled: boolean
  /** The updater is on, but the host agent that runs upgrades is not reporting. */
  agent_stopped?: boolean
  /** The current user may start an upgrade: the server's owner, while enabled. */
  can_upgrade: boolean
  /** A request is queued and the server has not picked it up yet. */
  pending: boolean
  pending_id?: string
  /** The current or last run. Empty when there has never been one. */
  id: string
  /** running, succeeded, failed, interrupted, or empty. */
  state: string
  step: string
  channel: string
  cli_from: string
  cli_to: string
  started_at: string
  finished_at: string
  error: string
  /** "enterprise" when the run switches to the Enterprise images. */
  edition?: string
  /** The end of the run's log. Only the server's owner receives it. */
  log_tail: string[]
}

export type ReleaseChannel = "stable" | "edge"

export interface ChannelCommit {
  /** Short commit. */
  sha: string
  /** First line of the commit message. Absent when only the commit is known. */
  subject: string
  date: string
}

/**
 * Both release channels, where this server is on them, and whether it may move
 * to the other one. Built from GitHub; `unavailable` says what is missing when
 * it could not be reached.
 */
export interface Channels {
  current: {
    version: string
    /** stable, edge, or empty for a development build. */
    channel: string
    /** The commit an edge build was cut from. */
    commit: string
  }
  stable: { releases: { tag: string; published_at: string; url: string }[] }
  edge: {
    /** The newest commit an edge server can pull. */
    head?: ChannelCommit
    /** Commits on main since the latest release. */
    ahead_by: number
    /** Those commits, newest first, at most 30. */
    commits: ChannelCommit[]
  }
  /** Absent on a development build. */
  switch?: {
    to: ReleaseChannel
    /** Only forward: edge to stable waits until a release includes the build. */
    allowed: boolean
    reason?: string
    /** What the switch brings, newest first, at most 30 of `total`. */
    changes: ChannelCommit[]
    total: number
  }
  unavailable?: string
}

export const system = {
  versionInfo: (token: string) =>
    apiFetch<VersionInfo>("/api/v1/system/version", {}, token),

  /** Ask GitHub now, past the API's caches. At most one real check every 30 seconds. */
  checkForUpdates: (token: string) =>
    apiFetch<VersionInfo>("/api/v1/system/check-updates", { method: "POST" }, token),

  exposure: (token: string) =>
    apiFetch<Exposure>("/api/v1/system/exposure", {}, token),

  /** Every advisory the current user has dismissed. One call for all of them. */
  dismissedNotices: (token: string) =>
    apiFetch<{ dismissed: string[] }>("/api/v1/system/notices", {}, token),

  dismissNotice: (key: string, token: string) =>
    apiFetch<void>(`/api/v1/system/notices/${key}/dismiss`, { method: "POST" }, token),

  upgradeStatus: (token: string) =>
    apiFetch<UpgradeStatus>("/api/v1/system/upgrade", {}, token),

  /**
   * Upgrade on the server's own channel and edition, or switch channel, or
   * switch to the Enterprise images the active licence grants. The image comes
   * from the licence on the server, never from here.
   */
  requestUpgrade: (token: string, opts?: { channel?: ReleaseChannel; edition?: "enterprise" }) =>
    apiFetch<UpgradeStatus>(
      "/api/v1/system/upgrade",
      opts && (opts.channel || opts.edition)
        ? { method: "POST", body: JSON.stringify(opts) }
        : { method: "POST" },
      token
    ),

  hostAgent: (token: string) =>
    apiFetch<HostAgentStatus>("/api/v1/system/host-agent", {}, token),

  /** Where migrating this server off another platform has got to. */
  migration: (token: string) =>
    apiFetch<MigrationState>("/api/v1/system/migrate/dokploy", {}, token),

  /**
   * Ask the host agent to run a stage. It is queued, not run here: the API has
   * no access to the host, and the agent on the gateway does the work.
   */
  requestMigration: (token: string, kind: MigrationStage, body?: { group?: string; volumes?: boolean }) =>
    apiFetch<HostRequestState>(
      `/api/v1/system/migrate/dokploy/${kind}`,
      { method: "POST", body: JSON.stringify(body ?? {}) },
      token
    ),

  channels: (token: string) =>
    apiFetch<Channels>("/api/v1/system/channels", {}, token),
}

File: packages/server/handler/system.go
package handler

import (
	"context"
	"errors"
	"net/http"

	"github.com/danielgtaylor/huma/v2"
	"github.com/meshploy/packages/server/service"
)

type VersionInfoOutput struct {
	Body *service.VersionInfo
}

func (h *Handler) registerSystemRoutes(api huma.API) {
	huma.Register(api, huma.Operation{
		OperationID: "get-version",
		Method:      "GET",
		Path:        "/api/v1/system/version",
		Summary:     "Get current and latest platform version",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetVersion)

	huma.Register(api, huma.Operation{
		OperationID: "get-host-agent",
		Method:      "GET",
		Path:        "/api/v1/system/host-agent",
		Summary:     "Whether the gateway's host agent is reporting",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetHostAgent)

	huma.Register(api, huma.Operation{
		OperationID: "get-dokploy-migration",
		Method:      "GET",
		Path:        "/api/v1/system/migrate/dokploy",
		Summary:     "The latest Dokploy detection and plan from the host agent, and their requests",
		Description: "Instance owner only: the plan names every app, domain and path on the server.",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetDokployMigration)

	huma.Register(api, huma.Operation{
		OperationID:   "request-dokploy-migration",
		Method:        "POST",
		Path:          "/api/v1/system/migrate/dokploy/{kind}",
		Summary:       "Ask the host agent to detect Dokploy or plan moving it",
		Description:   "Read-only on the host. Queues the request; follow it with GET /system/migrate/dokploy.",
		Tags:          []string{"System"},
		Security:      []map[string][]string{{"bearer": {}}},
		DefaultStatus: 202,
	}, h.RequestDokployMigration)

	huma.Register(api, huma.Operation{
		OperationID: "check-for-updates",
		Method:      "POST",
		Path:        "/api/v1/system/check-updates",
		Summary:     "Ask GitHub for the newest release or build now, past the caches",
		Description: "Forced checks are limited to one every 30 seconds; a sooner one returns the last answer.",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.CheckForUpdates)

	huma.Register(api, huma.Operation{
		OperationID: "get-exposure",
		Method:      "GET",
		Path:        "/api/v1/system/exposure",
		Summary:     "Report whether this gateway runs without a host firewall",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetExposure)

	huma.Register(api, huma.Operation{
		OperationID: "dismiss-notice",
		Method:      "POST",
		Path:        "/api/v1/system/notices/{key}/dismiss",
		Summary:     "Dismiss a console advisory for the current user",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.DismissNotice)

	huma.Register(api, huma.Operation{
		OperationID: "get-upgrade",
		Method:      "GET",
		Path:        "/api/v1/system/upgrade",
		Summary:     "Report whether this server can be upgraded from the console, and the last upgrade",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetUpgrade)

	huma.Register(api, huma.Operation{
		OperationID:   "request-upgrade",
		Method:        "POST",
		Path:          "/api/v1/system/upgrade",
		Summary:       "Queue an upgrade of this server to the latest build on its channel, or a switch to the other channel",
		Tags:          []string{"System"},
		Security:      []map[string][]string{{"bearer": {}}},
		DefaultStatus: http.StatusAccepted,
	}, h.RequestUpgrade)

	huma.Register(api, huma.Operation{
		OperationID: "get-channels",
		Method:      "GET",
		Path:        "/api/v1/system/channels",
		Summary:     "Describe the stable and edge channels, where this server is, and whether it may switch",
		Tags:        []string{"System"},
		Security:    []map[string][]string{{"bearer": {}}},
	}, h.GetChannels)
}

type ExposureOutput struct {
	Body *service.Exposure
}

// GetExposure is deliberately readable by any authenticated member rather than
// admins only: it reports nothing an operator could act on maliciously -- the
// ports are Meshploy's own published defaults, documented in the compose file --
// and hiding it from the person who happens to be signed in is how it goes
// unnoticed.
func (h *Handler) GetExposure(ctx context.Context, _ *struct{}) (*ExposureOutput, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	out, err := h.svc.System.GetExposure(ctx, userID)
	if err != nil {
		return nil, huma.Error500InternalServerError("failed to read exposure state", err)
	}
	return &ExposureOutput{Body: &out}, nil
}

type DismissNoticeInput struct {
	Key string `path:"key" doc:"Notice key, e.g. host-exposure"`
}

func (h *Handler) DismissNotice(ctx context.Context, input *DismissNoticeInput) (*struct{}, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	if err := h.svc.System.DismissNotice(ctx, userID, input.Key); err != nil {
		return nil, huma.Error400BadRequest("could not dismiss notice", err)
	}
	return nil, nil
}

func (h *Handler) GetVersion(ctx context.Context, _ *struct{}) (*VersionInfoOutput, error) {
	if _, err := requireUser(ctx); err != nil {
		return nil, err
	}
	info := h.svc.System.GetVersionInfo(ctx)
	return &VersionInfoOutput{Body: &info}, nil
}

func (h *Handler) GetDokployMigration(ctx context.Context, _ *struct{}) (*struct{ Body service.MigrationState }, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	st, err := h.svc.System.GetMigrationState(ctx, userID)
	if errors.Is(err, service.ErrMigrationNotOwner) {
		return nil, huma.Error403Forbidden(err.Error())
	}
	if err != nil {
		return nil, err
	}
	return &struct{ Body service.MigrationState }{Body: st}, nil
}

func (h *Handler) RequestDokployMigration(ctx context.Context, in *struct {
	Kind string `path:"kind" enum:"detect,plan"`
}) (*struct{ Body service.HostRequestState }, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	st, err := h.svc.System.RequestMigration(ctx, userID, in.Kind)
	switch {
	case errors.Is(err, service.ErrMigrationNotOwner):
		return nil, huma.Error403Forbidden(err.Error())
	case errors.Is(err, service.ErrHostAgentNotReporting):
		return nil, huma.Error409Conflict(err.Error())
	case errors.Is(err, service.ErrUnknownHostRequest):
		return nil, huma.Error404NotFound(err.Error())
	case err != nil:
		return nil, err
	}
	return &struct{ Body service.HostRequestState }{Body: st}, nil
}

func (h *Handler) GetHostAgent(ctx context.Context, _ *struct{}) (*struct{ Body service.HostAgentStatus }, error) {
	if _, err := requireUser(ctx); err != nil {
		return nil, err
	}
	return &struct{ Body service.HostAgentStatus }{Body: h.svc.System.HostAgentStatus()}, nil
}

func (h *Handler) CheckForUpdates(ctx context.Context, _ *struct{}) (*VersionInfoOutput, error) {
	if _, err := requireUser(ctx); err != nil {
		return nil, err
	}
	info := h.svc.System.CheckForUpdates(ctx)
	return &VersionInfoOutput{Body: &info}, nil
}

type UpgradeStatusOutput struct {
	Body *service.UpgradeStatus
}

// GetUpgrade is readable by any member, so everyone signed in can see that an
// upgrade is under way; the service gives the log only to the instance owner.
func (h *Handler) GetUpgrade(ctx context.Context, _ *struct{}) (*UpgradeStatusOutput, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	st, err := h.svc.System.GetUpgradeStatus(ctx, userID)
	if err != nil {
		return nil, huma.Error500InternalServerError("failed to read the upgrade status", err)
	}
	return &UpgradeStatusOutput{Body: &st}, nil
}

type RequestUpgradeInput struct {
	// Optional: a bare POST upgrades on the server's own channel.
	Body *struct {
		Channel string `json:"channel,omitempty" enum:"stable,edge" doc:"Switch to this channel. Omit to stay on the server's own."`
		Edition string `json:"edition,omitempty" enum:"enterprise" doc:"Switch to the Enterprise images the active licence grants. Omit to keep the images the server runs."`
	}
}

// RequestUpgrade upgrades on the server's own channel, or switches channel or
// edition. What reaches the host is the channel, which the runner checks again,
// and for a switch to Enterprise the image the verified licence names, never
// one the client sent.
func (h *Handler) RequestUpgrade(ctx context.Context, input *RequestUpgradeInput) (*UpgradeStatusOutput, error) {
	userID, err := requireUser(ctx)
	if err != nil {
		return nil, err
	}
	var opts service.UpgradeOptions
	if input.Body != nil {
		opts.Channel, opts.Edition = input.Body.Channel, input.Body.Edition
	}
	st, err := h.svc.System.RequestUpgrade(ctx, userID, opts)
	switch {
	case errors.Is(err, service.ErrNotInstanceOwner):
		return nil, huma.Error403Forbidden(err.Error())
	case errors.Is(err, service.ErrUnknownChannel), errors.Is(err, service.ErrUnknownEdition):
		return nil, huma.Error400BadRequest(err.Error())
	case errors.Is(err, service.ErrUpgradeNotEnabled),
		errors.Is(err, service.ErrUpgradeRunning),
		errors.Is(err, service.ErrUpgradeDevBuild),
		errors.Is(err, service.ErrChannelSwitchRefused),
		errors.Is(err, service.ErrEditionSwitchRefused):
		return nil, huma.Error409Conflict(err.Error())
	case err != nil:
		return nil, huma.Error500InternalServerError("failed to queue the upgrade", err)
	}
	return &UpgradeStatusOutput{Body: &st}, nil
}

type ChannelsOutput struct {
	Body *service.Channels
}

// GetChannels is readable by any member, like the version; switching is the
// instance owner's, and RequestUpgrade checks that.
func (h *Handler) GetChannels(ctx context.Context, _ *struct{}) (*ChannelsOutput, error) {
	if _, err := requireUser(ctx); err != nil {
		return nil, err
	}
	out := h.svc.System.GetChannels(ctx)
	return &ChannelsOutput{Body: &out}, nil
}
